### ISSUE TITLE
[1.19.2] IC2Classic StackCrafting support

### ISSUE THREAD

**Issue by CrossVas** [emilyploszaj/emi#352]
Currently EMI doesn't work well with IC2Classic's StackCrafting recipes. 
## Actual recipe
*displayed by JEI & REI 
![image](https://github.com/emilyploszaj/emi/assets/25727222/2400831e-7f9f-4619-b6d9-d4b5c32a737c)

## Recipe shown by EMI
![image](https://github.com/emilyploszaj/emi/assets/25727222/98255200-8c07-4ca1-b29d-5e50daa68f84)

**Comment by Flamarine** [comment]
this should be done on the ic2c side rather than the emi side ig?

**Comment by CrossVas** [comment]
> this should be done on the ic2c side rather than the emi side ig?

Didn't mention from the beginning, but the upper image is from JEI & REI, they work just fine (I hate comparing mods, but I have no choice in this case). 

EDIT: 
And afaik they didn't do anything special in order to handle these recipes nor did the IC2Classic. They just work.

**Comment by Flamarine** [comment]
> > this should be done on the ic2c side rather than the emi side ig?
> 
> Didn't mention from the beginning, but the upper image is from JEI & REI, they work just fine (I hate comparing mods, but I have no choice in this case). 
> 
> EDIT: 
> And afaik they didn't do anything special in order to handle these recipes nor did the IC2Classic. They just work.

it's ic2c itself that adds compat, not jei or rei

**Comment by CrossVas** [comment]
> it's ic2c itself that adds compat, not jei or rei

IC2Classic adds 0 compat for JEI/REI on this. JEI/REI work out of the box.
I was just wondering if the main dev could take a look on how they do it and add support too. It would be much appreciated.

**Comment by Flamarine** [comment]
> > it's ic2c itself that adds compat, not jei or rei
> 
> IC2Classic adds 0 compat for JEI/REI on this. JEI/REI work out of the box. I was just wondering if the main dev could take a look on how they do it and add support too. It would be much appreciated.

it's **builtin** in **that** mod and since IC2C is closed source, you cannot actually see it

**Comment by Speiger** [comment]
o/ @Flamarine IC2Classic dev here.

First of all the API you are talking about is open source on the IC2Classic repo, and that is all you need.   
Because we have to return "ingredients" at some point anyways.   

And i assume you use the "ingredients" from the Recipe class for displaying the items that should be used in the input.
Fun fact: Ingredients can have ItemStack instances, and JEI/REI simply use these ItemStack instances and AT MOST copy them.

That means any Ingredient they use automatically supports stacksize visualization.
In other words you have special code in your recipe viewer that removes the stacksize of the item or you are using a reverse search to see which items are part of the ingredient instead of using the (Mojangmaps 1.19.2) Ingredient#getItems function.

So yeah this is not a IC2Classic issue but a EMI issue.

And fun fact even if you disable the JEI plugin in IC2Classic, JEI stuff still works fine even with stacksize because the Recipe class from Mojang has stacksize support inferred.

Anyways, its a legit bugreport.

**Comment by Flamarine** [comment]
> o/ @Flamarine IC2Classic dev here.
> 
> First of all the API you are talking about is open source on the IC2Classic repo, and that is all you need.   
> Because we have to return "ingredients" at some point anyways.   
> 
> And i assume you use the "ingredients" from the Recipe class for displaying the items that should be used in the input.
> Fun fact: Ingredients can have ItemStack instances, and JEI/REI simply use these ItemStack instances and AT MOST copy them.
> 
> That means any Ingredient they use automatically supports stacksize visualization.
> In other words you have special code in your recipe viewer that removes the stacksize of the item or you are using a reverse search to see which items are part of the ingredient instead of using the (Mojangmaps 1.19.2) Ingredient#getItems function.
> 
> So yeah this is not a IC2Classic issue but a EMI issue.
> 
> And fun fact even if you disable the JEI plugin in IC2Classic, JEI stuff still works fine even with stacksize because the Recipe class from Mojang has stacksize support inferred.
> 
> Anyways, its a legit bugreport.

wow, got it.

**Comment by Speiger** [comment]
Also i am toying with the idea of making a open source implementation out of stacksize recipes anyways since i got so many requests.

So this will be a good fix before you get rushed reports from everywhere.

**Comment by emilyploszaj** [comment]
This is related to an ingredient parsing error. Now if all stacks in an ingredient have the same size, it'll use that amount instead of 1 as of 1.0.28

**Comment by Speiger** [comment]
@emilyploszaj thank you for the fix!
